pysys/process/monitorimpl.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python
+# PySys System Test Framework, Copyright (C) 2006-2019 M.B. Grieve
+
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+"""
+Contains implementations of the L{BaseProcessMonitor} interface. 
+"""
+
+__all__ = ['DEFAULT_PROCESS_MONITOR', 'WindowsProcessMonitor', 'UnixProcessMonitor']
+
+
+import os, sys, string, time, logging
+from pysys import process_lock
+from pysys.constants import *
+from pysys.utils.pycompat import *
+from pysys.process.monitor import *
+
+if PLATFORM=='win32' and 'epydoc' not in sys.modules:
+	import win32api, win32pdh, win32con, win32process
+
+log = logging.getLogger('pysys.processmonitor')
+
+class WindowsProcessMonitor(BaseProcessMonitor):
+	"""
+	Windows implementation of a process monitor. 
+	
+	Uses the `GetProcessMemoryInfo`, and `GetProcessTimes` APIs. 
+	The UserTime and KernelTime are summed together to calculate the CPU 
+	utilization for this process. 
+	"""
+	
+	def start(self):
+		self._hPid = win32api.OpenProcess(win32con.PROCESS_QUERY_INFORMATION, 0, self.pid)
+		self._lastValues = None
+
+		return BaseProcessMonitor.start(self)
+
+	def _timer_ns(self):
+		""" Return a monotonically incrementing clock value in nanoseconds (1000**3) that 
+		can be used to calculate performance results. 
+		"""
+		return win32api.GetTickCount()*1000000 # millis->nanos.
+
+	def _cleanup(self):
+		self._hPid.close()
+
+	def _getData(self, sample):
+		while True: # loop until we have both a "new" and a "last" value for CPU time
+			if self._stopping.isSet(): raise Exception('Requested to stop')
+
+			newvalues = {}
+			newvalues['time_ns'] = self._timer_ns()
+			cputimes = win32process.GetProcessTimes(self._hPid)
+			newvalues['cputime_ns'] = (cputimes['KernelTime']+cputimes['UserTime'])*100 # convert to ns; comes in 100*ns units
+
+			if self._lastValues is not None:
+				if newvalues['time_ns']-self._lastValues['time_ns'] <= 0:
+					# wait a bit longer to avoid div by zero error if the sleeping is somehow messed up
+					self._stopping.wait(min(self.interval, 1))
+					continue
+			
+				lastvalues = self._lastValues
+				break
+			# this is just for the first time _getData is called; need to repeat this once so we have stats to compare to
```

Comment:
Uber-nitpick: add a blank line before the comment :-)
```suggestion
			
			# this is just for the first time _getData is called; need to repeat this once so we have stats to compare to
```

Reply:
fixed